jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/Booster.java

```diff
@@ -316,6 +316,159 @@ private synchronized float[][] predict(DMatrix data,
+   * <code>inplace_predict(data, num_rows, num_features, missing,
+   *                       outputMargin, treeLimit, false, false)</code>.
+   *
+   * @param data           Flattened input matrix of features for prediction
+   * @param num_rows       The number of preditions to make (count of input matrix rows)
+   * @param num_features   The number of features in the model (count of input matrix columns)
+   * @param missing        Value indicating missing element in the <code>data</code> input matrix
+   * @param outputMargin   Whether to only predict margin value instead of transformed prediction
+   * @param treeLimit      limit number of trees, 0 means all trees.
+   *
+   * @return predict       Result matrix
+   *
+   * @see #inplace_predict(float[] data, int num_rows, int num_features, float missing,
+   *                       boolean outputMargin, int treeLimit, boolean predLeaf,
+   *                       boolean predContribs)
+   */
+  public float[][] inplace_predict(float[] data,
+                                   int num_rows,
+                                   int num_features,
+                                   float missing,
+                                   boolean outputMargin,
+                                   int treeLimit) throws XGBoostError {
+    return this.inplace_predict(data, num_rows, num_features, missing,
+                                outputMargin, treeLimit, false, false);
+  }
+
+  /**
+   * Perform thread-safe prediction.
+   *
+   * @param data           Flattened input matrix of features for prediction
+   * @param num_rows       The number of preditions to make (count of input matrix rows)
+   * @param num_features   The number of features in the model (count of input matrix columns)
+   * @param d_matrix_h     The handle for a dmatrix
+   * @param missing        Value indicating missing element in the <code>data</code> input matrix
+   * @param outputMargin   Whether to only predict margin value instead of transformed prediction
+   * @param treeLimit      limit number of trees, 0 means all trees.
+   * @param predLeaf       prediction minimum to keep leafs
+   * @param predContribs   prediction feature contributions
+   *
+   * @return predict       Result matrix
+   */
+  public float[][] inplace_predict(float[] data,
+                                   int num_rows,
+                                   int num_features,
+                                   float missing,
+                                   boolean outputMargin,
+                                   int treeLimit,
+                                   boolean predLeaf,
+                                   boolean predContribs) throws XGBoostError {
+    int optionMask = 0;
+    if (outputMargin) {
+      optionMask = 1;
+    }
+    if (predLeaf) {
+      optionMask = 2;
+    }
+    if (predContribs) {
+      optionMask = 4;
+    }
+    DMatrix d_mat = new DMatrix(data, num_rows, num_features, missing);
```

Comment:
Hmm, `inplace_predict` doesn't need the `DMatrix`, which is the primary motivation of developing this method.

Reply:
Sorry for the delay - I'm looking through the python code that calls predictFromDense to see how that works - should we be implementing and using proxy DMatrix on the Java side of things? I'll start looking at making that change if you can just let me know if that would be the correct approach. 

Historically, we didn't have DMatrix here when we initially made our changes in 1.4 - this addition of the DMatrix here was due to changes in `cpu_predictor.cc` that now requires a DMatrix, we were originally using the parameter `dmlc::any const &x` in the old version of inplacePredict there and sending a null pointer for the DMatrix

Reply:
Thank you for sharing! We unified the interface under the proxy DMatrix class. Let me do some more investigation into the JVM implementation first. I think I can help with the C++ code.